File: backend/src/main/java/br/com/calciolari/datahub/imports/infrastructure/interpdv/qrp/InterPdvReportLayoutMapper.java
package br.com.calciolari.datahub.imports.infrastructure.interpdv.qrp;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import br.com.calciolari.datahub.imports.domain.parser.IssueSeverity;
import br.com.calciolari.datahub.imports.domain.parser.IssueStage;
import br.com.calciolari.datahub.imports.domain.parser.MovementDirection;
import br.com.calciolari.datahub.imports.domain.parser.ParseIssue;
import br.com.calciolari.datahub.imports.domain.parser.ParsedImport;
import br.com.calciolari.datahub.imports.domain.parser.ParsedImportStats;
import br.com.calciolari.datahub.imports.domain.parser.ParsedImportTotals;
import br.com.calciolari.datahub.imports.domain.parser.ParsedMovement;
import br.com.calciolari.datahub.imports.domain.parser.SourceLocator;

/**
 * Maps EMF text runs onto the InterPDV "Relatório de Auditoria" layout.
 * Port of PoC {@code parseQrp} column association (x-distance &lt; 75).
 */
public final class InterPdvReportLayoutMapper {

	public static final String SOURCE = "INTERPDV";
	public static final String PARSER_NAME = "interpdv-qrp";
	public static final String PARSER_VERSION = "interpdv-qrp-v1";

	private static final Pattern PRODUCT = Pattern.compile("^Produto:\\s*(\\d+)\\s*-\\s*(.+)$", Pattern.CASE_INSENSITIVE);
	private static final Pattern SALE = Pattern.compile("^Venda Numero:\\s*(\\d+)$", Pattern.CASE_INSENSITIVE);
	private static final Pattern EXPORT_DT = Pattern.compile("^Data/Hora:\\s*(\\d{2}/\\d{2}/\\d{4})\\s+(\\d{2}:\\d{2}:\\d{2})$", Pattern.CASE_INSENSITIVE);
	private static final Pattern MANUFACTURER = Pattern.compile("^Fabricante:\\s*(.*)$", Pattern.CASE_INSENSITIVE);
	private static final Pattern STOCK = Pattern.compile("^Estoque Atual:\\s*(.*)$", Pattern.CASE_INSENSITIVE);

	private static final List<String> HEADERS = List.of(
			"Preço", "Desconto", "Total Item", "Data", "Hora", "Saidas", "Entradas", "Anterior", "Posterior");

	private static final int COLUMN_X_TOLERANCE = 75;
	private static final DateTimeFormatter BR_DATE = DateTimeFormatter.ofPattern("dd/MM/yyyy");
	private static final DateTimeFormatter BR_TIME = DateTimeFormatter.ofPattern("HH:mm:ss");

	private final int columnTolerance;

	public InterPdvReportLayoutMapper() {
		this(COLUMN_X_TOLERANCE);
	}

	public InterPdvReportLayoutMapper(int columnTolerance) {
		if (columnTolerance < 1) {
			throw new IllegalArgumentException("columnTolerance must be >= 1");
		}
		this.columnTolerance = columnTolerance;
	}

	public ParsedImport map(List<EmfTextRun> items, int pageCount) {
		Objects.requireNonNull(items, "items");
		List<ParseIssue> issues = new ArrayList<>();

		Optional<Matcher> product = firstMatch(items, PRODUCT);
		if (product.isEmpty()) {
			issues.add(issue("PRODUCT_FIELD_MISSING", IssueSeverity.FATAL, IssueStage.LAYOUT,
					"Campo Produto não encontrado no relatório."));
			return emptyFatal(issues, pageCount);
		}

		String externalProductId = product.get().group(1);
		String productName = product.get().group(2).trim();
		String manufacturer = firstMatch(items, MANUFACTURER).map(m -> m.group(1).trim()).orElse(null);

		Map<Integer, Map<String, Integer>> headerByPage = new HashMap<>();
		for (EmfTextRun it : items) {
			if (HEADERS.contains(it.text())) {
				headerByPage.computeIfAbsent(it.pageIndex(), p -> new HashMap<>()).put(it.text(), it.x());
			}
		}

		Map<String, List<EmfTextRun>> groups = new LinkedHashMap<>();
		for (EmfTextRun it : items) {
			String key = it.pageIndex() + ":" + it.y();
			groups.computeIfAbsent(key, k -> new ArrayList<>()).add(it);
		}

		List<ParsedMovement> movements = new ArrayList<>();
		int recordIndex = 0;
		for (List<EmfTextRun> group : groups.values()) {
			EmfTextRun sale = null;
			Matcher saleMatcher = null;
			for (EmfTextRun it : group) {
				Matcher m = SALE.matcher(it.text());
				if (m.matches()) {
					sale = it;
					saleMatcher = m;
					break;
				}
			}
			if (sale == null) {
				continue;
			}

			Map<String, Integer> hp = headerByPage.getOrDefault(sale.pageIndex(), Map.of());
			Map<String, String> row = new HashMap<>();
			row.put("saleNumber", saleMatcher.group(1));
			for (EmfTextRun it : group) {
				if (it == sale) {
					continue;
				}
				String best = null;
				int dist = Integer.MAX_VALUE;
				for (String header : HEADERS) {
					Integer hx = hp.get(header);
					if (hx == null) {
						continue;
					}
					int d = Math.abs(it.x() - hx);
					if (d < dist) {
						dist = d;
						best = header;
					}
				}
				if (best != null && dist < columnTolerance) {
					row.put(best, it.text());
				}
			}

			BigDecimal exits = BrazilianDecimalParser.parse(row.get("Saidas"));
			BigDecimal entries = BrazilianDecimalParser.parse(row.get("Entradas"));
			MovementDirection direction = resolveDirection(exits, entries);

			LocalDateTime occurredAt = parseOccurredAt(row.get("Data"), row.get("Hora"), issues, recordIndex);

			movements.add(new ParsedMovement(
					recordIndex,
					direction,
					externalProductId,
					productName,
					row.get("saleNumber"),
					occurredAt,
					quantityForDirection(direction, exits, entries),
					BrazilianDecimalParser.parse(row.get("Preço")),
					BrazilianDecimalParser.parse(row.get("Desconto")),
					BrazilianDecimalParser.parse(row.get("Total Item")),
					BrazilianDecimalParser.parse(row.get("Anterior")),
					BrazilianDecimalParser.parse(row.get("Posterior")),
					manufacturer,
					new SourceLocator(sale.pageIndex() + 1, recordIndex, null, "y=" + sale.y())));
			recordIndex++;
		}

		BigDecimal sourceQuantityTotal = findSourceQuantityTotal(items);
		BigDecimal parsedQuantityTotal = sumOutQuantities(movements);
		BigDecimal parsedRevenueTotal = sumOutTotals(movements);
		LocalDateTime first = movements.stream()
				.map(ParsedMovement::occurredAt)
				.filter(Objects::nonNull)
				.min(Comparator.naturalOrder())
				.orElse(null);
		LocalDateTime last = movements.stream()
				.map(ParsedMovement::occurredAt)
				.filter(Objects::nonNull)
				.max(Comparator.naturalOrder())
				.orElse(null);

		long uniqueSales = movements.stream()
				.map(ParsedMovement::externalSaleId)
				.filter(Objects::nonNull)
				.distinct()
				.count();
		int entryCount = (int) movements.stream().filter(m -> m.direction() == MovementDirection.IN).count();
		int exitCount = (int) movements.stream().filter(m -> m.direction() == MovementDirection.OUT).count();

		if (movements.isEmpty()) {
			issues.add(issue("NO_SALE_ROWS", IssueSeverity.ERROR, IssueStage.LAYOUT,
					"Nenhuma linha de venda reconhecida."));
		}

		return new ParsedImport(
				SOURCE,
				PARSER_NAME,
				PARSER_VERSION,
				externalProductId,
				productName,
				movements,
				new ParsedImportTotals(
						sourceQuantityTotal,
						parsedQuantityTotal,
						null,
						parsedRevenueTotal,
						first,
						last),
				new ParsedImportStats(pageCount, movements.size(), (int) uniqueSales, entryCount, exitCount),
				issues);
	}

	private static MovementDirection resolveDirection(BigDecimal exits, BigDecimal entries) {
		boolean hasExit = exits != null;
		boolean hasEntry = entries != null;
		if (hasExit && !hasEntry) {
			return MovementDirection.OUT;
		}
		if (hasEntry && !hasExit) {
			return MovementDirection.IN;
		}
		if (hasExit && hasEntry) {
			// PoC treats both columns; prefer explicit non-null exits for audit sales rows.
			return MovementDirection.OUT;
		}
		return MovementDirection.UNKNOWN;
	}

	private static BigDecimal quantityForDirection(MovementDirection direction, BigDecimal exits, BigDecimal entries) {
		return switch (direction) {
			case OUT -> exits;
			case IN -> entries;
			case RETURN, UNKNOWN -> exits != null ? exits : entries;
		};
	}

	private static BigDecimal sumOutQuantities(List<ParsedMovement> movements) {
		return movements.stream()
				.filter(m -> m.direction() == MovementDirection.OUT && m.quantity() != null)
				.map(ParsedMovement::quantity)
				.reduce(BigDecimal.ZERO, BigDecimal::add);
	}

	private static BigDecimal sumOutTotals(List<ParsedMovement> movements) {
		return movements.stream()
				.filter(m -> m.direction() == MovementDirection.OUT && m.total() != null)
				.map(ParsedMovement::total)
				.reduce(BigDecimal.ZERO, BigDecimal::add);
	}

	/**
	 * PoC/fixture evidence: label {@code Total de Vendas:} followed by a nearby numeric text run
	 * (observed as a separate EMF text on the same report footers).
	 */
	private static BigDecimal findSourceQuantityTotal(List<EmfTextRun> items) {
		for (int i = 0; i < items.size(); i++) {
			if ("Total de Vendas:".equalsIgnoreCase(items.get(i).text())) {
				for (int j = i + 1; j < Math.min(i + 6, items.size()); j++) {
					BigDecimal value = BrazilianDecimalParser.parse(items.get(j).text());
					if (value != null) {
						return value;
					}
				}
			}
		}
		return null;
	}

	private static LocalDateTime parseOccurredAt(String date, String time, List<ParseIssue> issues, int recordIndex) {
		if (date == null || date.isBlank() || time == null || time.isBlank()) {
			return null;
		}
		try {
			LocalDate d = LocalDate.parse(date.trim(), BR_DATE);
			LocalTime t = LocalTime.parse(time.trim(), BR_TIME);
			return LocalDateTime.of(d, t);
		}
		catch (DateTimeParseException ex) {
			issues.add(new ParseIssue(
					"INVALID_DATETIME",
					IssueSeverity.WARNING,
					IssueStage.MAPPING,
					new SourceLocator(null, recordIndex, null, null),
					"Data/hora inválida: " + date + " " + time));
			return null;
		}
	}

	private static Optional<Matcher> firstMatch(List<EmfTextRun> items, Pattern pattern) {
		for (EmfTextRun it : items) {
			Matcher m = pattern.matcher(it.text());
			if (m.matches()) {
				return Optional.of(m);
			}
		}
		return Optional.empty();
	}

	private static ParseIssue issue(String code, IssueSeverity severity, IssueStage stage, String message) {
		return new ParseIssue(code, severity, stage, SourceLocator.empty(), message);
	}

	private static ParsedImport emptyFatal(List<ParseIssue> issues, int pageCount) {
		return new ParsedImport(
				SOURCE,
				PARSER_NAME,
				PARSER_VERSION,
				null,
				null,
				List.of(),
				ParsedImportTotals.empty(),
				new ParsedImportStats(pageCount, 0, 0, 0, 0),
				issues);
	}
}

File: backend/src/Calciolari.DataHub/Imports/Infrastructure/InterPdv/Qrp/InterPdvReportLayoutMapper.cs
using System.Globalization;
using System.Text.RegularExpressions;
using Calciolari.DataHub.Imports.Domain.Parser;

namespace Calciolari.DataHub.Imports.Infrastructure.InterPdv.Qrp;

/// <summary>
/// Maps EMF text runs onto the InterPDV "Relatório de Auditoria" layout.
/// Port of PoC parseQrp column association (x-distance &lt; 75).
/// </summary>
public sealed class InterPdvReportLayoutMapper
{
    public const string Source = "INTERPDV";
    public const string ParserName = "interpdv-qrp";
    public const string ParserVersion = "interpdv-qrp-v1";

    private static readonly Regex Product = new(@"^Produto:\s*(\d+)\s*-\s*(.+)$", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex Sale = new(@"^Venda Numero:\s*(\d+)$", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex Manufacturer = new(@"^Fabricante:\s*(.*)$", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly string[] Headers =
    [
        "Preço", "Desconto", "Total Item", "Data", "Hora", "Saidas", "Entradas", "Anterior", "Posterior"
    ];

    private const int ColumnXTolerance = 75;
    private static readonly CultureInfo PtBr = CultureInfo.GetCultureInfo("pt-BR");

    private readonly int _columnTolerance;

    public InterPdvReportLayoutMapper()
        : this(ColumnXTolerance)
    {
    }

    public InterPdvReportLayoutMapper(int columnTolerance)
    {
        if (columnTolerance < 1)
        {
            throw new ArgumentOutOfRangeException(nameof(columnTolerance), "columnTolerance must be >= 1");
        }

        _columnTolerance = columnTolerance;
    }

    public ParsedImport Map(IReadOnlyList<EmfTextRun> items, int pageCount)
    {
        ArgumentNullException.ThrowIfNull(items);
        var issues = new List<ParseIssue>();

        var product = FirstMatch(items, Product);
        if (product is null)
        {
            issues.Add(Issue("PRODUCT_FIELD_MISSING", IssueSeverity.Fatal, IssueStage.Layout,
                "Campo Produto não encontrado no relatório."));
            return EmptyFatal(issues, pageCount);
        }

        var externalProductId = product.Groups[1].Value;
        var productName = product.Groups[2].Value.Trim();
        var manufacturer = FirstMatch(items, Manufacturer)?.Groups[1].Value.Trim();

        var headerByPage = new Dictionary<int, Dictionary<string, int>>();
        foreach (var it in items)
        {
            if (Headers.Contains(it.Text, StringComparer.Ordinal))
            {
                if (!headerByPage.TryGetValue(it.PageIndex, out var map))
                {
                    map = new Dictionary<string, int>(StringComparer.Ordinal);
                    headerByPage[it.PageIndex] = map;
                }

                map[it.Text] = it.X;
            }
        }

        var groups = new Dictionary<string, List<EmfTextRun>>();
        foreach (var it in items)
        {
            var key = it.PageIndex + ":" + it.Y;
            if (!groups.TryGetValue(key, out var group))
            {
                group = [];
                groups[key] = group;
            }

            group.Add(it);
        }

        var movements = new List<ParsedMovement>();
        var recordIndex = 0;
        foreach (var group in groups.Values)
        {
            EmfTextRun? sale = null;
            Match? saleMatcher = null;
            foreach (var it in group)
            {
                var m = Sale.Match(it.Text);
                if (m.Success)
                {
                    sale = it;
                    saleMatcher = m;
                    break;
                }
            }

            if (sale is null || saleMatcher is null)
            {
                continue;
            }

            headerByPage.TryGetValue(sale.PageIndex, out var hp);
            hp ??= new Dictionary<string, int>(StringComparer.Ordinal);
            var row = new Dictionary<string, string>(StringComparer.Ordinal)
            {
                ["saleNumber"] = saleMatcher.Groups[1].Value
            };
            foreach (var it in group)
            {
                if (ReferenceEquals(it, sale))
                {
                    continue;
                }

                string? best = null;
                var dist = int.MaxValue;
                foreach (var header in Headers)
                {
                    if (!hp.TryGetValue(header, out var hx))
                    {
                        continue;
                    }

                    var d = Math.Abs(it.X - hx);
                    if (d < dist)
                    {
                        dist = d;
                        best = header;
                    }
                }

                if (best is not null && dist < _columnTolerance)
                {
                    row[best] = it.Text;
                }
            }

            row.TryGetValue("Saidas", out var saidas);
            row.TryGetValue("Entradas", out var entradas);
            var exits = BrazilianDecimalParser.Parse(saidas);
            var entries = BrazilianDecimalParser.Parse(entradas);
            var direction = ResolveDirection(exits, entries);

            row.TryGetValue("Data", out var date);
            row.TryGetValue("Hora", out var time);
            var occurredAt = ParseOccurredAt(date, time, issues, recordIndex);

            row.TryGetValue("Preço", out var preco);
            row.TryGetValue("Desconto", out var desconto);
            row.TryGetValue("Total Item", out var totalItem);
            row.TryGetValue("Anterior", out var anterior);
            row.TryGetValue("Posterior", out var posterior);

            movements.Add(new ParsedMovement(
                recordIndex,
                direction,
                externalProductId,
                productName,
                row["saleNumber"],
                occurredAt,
                QuantityForDirection(direction, exits, entries),
                BrazilianDecimalParser.Parse(preco),
                BrazilianDecimalParser.Parse(desconto),
                BrazilianDecimalParser.Parse(totalItem),
                BrazilianDecimalParser.Parse(anterior),
                BrazilianDecimalParser.Parse(posterior),
                manufacturer,
                new SourceLocator(sale.PageIndex + 1, recordIndex, null, "y=" + sale.Y)));
            recordIndex++;
        }

        var sourceQuantityTotal = FindSourceQuantityTotal(items);
        var parsedQuantityTotal = SumOutQuantities(movements);
        var parsedRevenueTotal = SumOutTotals(movements);
        DateTime? first = movements
            .Select(m => m.OccurredAt)
            .Where(v => v is not null)
            .Min();
        DateTime? last = movements
            .Select(m => m.OccurredAt)
            .Where(v => v is not null)
            .Max();

        var uniqueSales = movements
            .Select(m => m.ExternalSaleId)
            .Where(id => id is not null)
            .Distinct()
            .Count();
        var entryCount = movements.Count(m => m.Direction == MovementDirection.In);
        var exitCount = movements.Count(m => m.Direction == MovementDirection.Out);

        if (movements.Count == 0)
        {
            issues.Add(Issue("NO_SALE_ROWS", IssueSeverity.Error, IssueStage.Layout,
                "Nenhuma linha de venda reconhecida."));
        }

        return new ParsedImport(
            Source,
            ParserName,
            ParserVersion,
            externalProductId,
            productName,
            movements,
            new ParsedImportTotals(
                sourceQuantityTotal,
                parsedQuantityTotal,
                null,
                parsedRevenueTotal,
                first,
                last),
            new ParsedImportStats(pageCount, movements.Count, uniqueSales, entryCount, exitCount),
            issues);
    }

    private static MovementDirection ResolveDirection(decimal? exits, decimal? entries)
    {
        var hasExit = exits is not null;
        var hasEntry = entries is not null;
        if (hasExit && !hasEntry)
        {
            return MovementDirection.Out;
        }

        if (hasEntry && !hasExit)
        {
            return MovementDirection.In;
        }

        if (hasExit && hasEntry)
        {
            return MovementDirection.Out;
        }

        return MovementDirection.Unknown;
    }

    private static decimal? QuantityForDirection(MovementDirection direction, decimal? exits, decimal? entries) =>
        direction switch
        {
            MovementDirection.Out => exits,
            MovementDirection.In => entries,
            MovementDirection.Return or MovementDirection.Unknown => exits ?? entries,
            _ => exits ?? entries
        };

    private static decimal SumOutQuantities(IEnumerable<ParsedMovement> movements) =>
        movements
            .Where(m => m.Direction == MovementDirection.Out && m.Quantity is not null)
            .Select(m => m.Quantity!.Value)
            .DefaultIfEmpty(0m)
            .Sum();

    private static decimal SumOutTotals(IEnumerable<ParsedMovement> movements) =>
        movements
            .Where(m => m.Direction == MovementDirection.Out && m.Total is not null)
            .Select(m => m.Total!.Value)
            .DefaultIfEmpty(0m)
            .Sum();

    private static decimal? FindSourceQuantityTotal(IReadOnlyList<EmfTextRun> items)
    {
        for (var i = 0; i < items.Count; i++)
        {
            if (string.Equals(items[i].Text, "Total de Vendas:", StringComparison.OrdinalIgnoreCase))
            {
                var limit = Math.Min(i + 6, items.Count);
                for (var j = i + 1; j < limit; j++)
                {
                    var value = BrazilianDecimalParser.Parse(items[j].Text);
                    if (value is not null)
                    {
                        return value;
                    }
                }
            }
        }

        return null;
    }

    private static DateTime? ParseOccurredAt(string? date, string? time, List<ParseIssue> issues, int recordIndex)
    {
        if (string.IsNullOrWhiteSpace(date) || string.IsNullOrWhiteSpace(time))
        {
            return null;
        }

        if (DateTime.TryParseExact(
                date.Trim() + " " + time.Trim(),
                "dd/MM/yyyy HH:mm:ss",
                PtBr,
                DateTimeStyles.None,
                out var parsed))
        {
            return DateTime.SpecifyKind(parsed, DateTimeKind.Unspecified);
        }

        issues.Add(new ParseIssue(
            "INVALID_DATETIME",
            IssueSeverity.Warning,
            IssueStage.Mapping,
            new SourceLocator(null, recordIndex, null, null),
            "Data/hora inválida: " + date + " " + time));
        return null;
    }

    private static Match? FirstMatch(IReadOnlyList<EmfTextRun> items, Regex pattern)
    {
        foreach (var it in items)
        {
            var m = pattern.Match(it.Text);
            if (m.Success)
            {
                return m;
            }
        }

        return null;
    }

    private static ParseIssue Issue(string code, IssueSeverity severity, IssueStage stage, string message) =>
        new(code, severity, stage, SourceLocator.Empty, message);

    private static ParsedImport EmptyFatal(List<ParseIssue> issues, int pageCount) =>
        new(
            Source,
            ParserName,
            ParserVersion,
            null,
            null,
            Array.Empty<ParsedMovement>(),
            ParsedImportTotals.Empty,
            new ParsedImportStats(pageCount, 0, 0, 0, 0),
            issues);
}
